### PR TITLE
[HOPSWORKS-1177] Allow users to manage 3rd party API keys

### PR DESCRIPTION
**Pull request by kouzant**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [ ] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [ ] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**
https://logicalclocks.atlassian.net/browse/HOPSWORKS-1177

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
New feature

* **What is the new behavior (if this is a feature change)?**
Users, from their accounts settings will be able to manage 3rd party API keys

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/user/UsersResource.java [hunk 1]
```diff
@@ -192,6 +200,42 @@ public Response changeLoginCredentials(
     return Response.ok().entity(json).build();
   }
 
+  @POST
+  @Path("apiKey")
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/user/UsersResource.java]: @kouzant this should be plural, like `apiKeys`. Same for `GET`

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/user/UsersResource.java [hunk 2]
```diff
@@ -192,6 +200,42 @@ public Response changeLoginCredentials(
...
+    List<ThirdPartyApiKeyPlaintext> apiKeys = thirdPartyApiKeysController.getAllApiKeysForUser(user);
+    ThirdPartyApiKeyDTO dto = thirdPartyApiKeysBuilder.build(apiKeys);
+    return Response.ok().entity(dto).build();
+  }
+  
+  @DELETE
+  @Path("apiKey/{keyName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Deletes an API key by its name")
+  public Response deleteApiKey(@PathParam("keyName") String keyName, @Context SecurityContext sc)
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/user/UsersResource.java]: @kouzant how about having an endpoint that deletes all the apiKeys?

File: hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java [hunk 3]
```diff
@@ -831,7 +831,11 @@ public int getRange() {
     TWO_FA_DISABLED(45, "2-factor authentication is disabled.", Response.Status.PRECONDITION_FAILED),
     TRANSITION_STATUS_ERROR(46, "The user can't transition from current status to requested status",
       Response.Status.BAD_REQUEST),
-    ACCESS_CONTROL(47, "Client not authorized for this invocation.", Response.Status.FORBIDDEN);
+    ACCESS_CONTROL(47, "Client not authorized for this invocation.", Response.Status.FORBIDDEN),
+    THIRD_PARTY_API_KEY_EMPTY(48, "API key is empty", Response.Status.BAD_REQUEST),
+    THIRD_PARTY_API_KEY_EXISTS(49, "Same API key already exists", Response.Status.BAD_REQUEST),
```
Comment (tkakantousis) [inline hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java]: @kouzant I'd know for HTTP 409 instead

Comment (kouzant) [reply hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java]: For `THIRD_PARTY_API_KEY_EXISTS` only or for both?

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/user/UsersResource.java (hunk 2 above)
Comment (kouzant) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/user/UsersResource.java]: I could add one to delete all API keys for user

File: hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java (hunk 3 above)
Comment (tkakantousis) [reply hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java]: `THIRD_PARTY_API_KEY_EXISTS` should be 409, `THIRD_PARTY_API_KEY_EMPTY` sounds like 404?

Comment (kouzant) [reply hopsworks-rest-utils/src/main/java/io/hops/hopsworks/restutils/RESTCodes.java]: Sure, it seems better

**Comment by kouzant** [comment]
New PR https://github.com/logicalclocks/hopsworks/pull/282

**Pull request closed without merging**